src/Shared/UrlDecoder/UrlDecoder.cs

```diff
@@ -607,33 +588,48 @@ private static int UnescapePercentEncoding(ref int scan, ReadOnlySpan<char> buff
         /// <returns>The hexadecimal value if successes, otherwise -1.</returns>
         private static int ReadHex(ref int scan, ReadOnlySpan<char> buffer)
         {
-            if (scan == buffer.Length)
+            // To eliminate boundary checks, using a temporary variable tempIdx.
+            int tempIdx = scan;
+            if ((uint)tempIdx >= (uint)buffer.Length)
             {
                 return -1;
             }
+            int value = buffer[tempIdx];
+            scan = tempIdx + 1;
 
-            var value = buffer[scan++];
-            var isHex = ((value >= '0') && (value <= '9')) ||
-                         ((value >= 'A') && (value <= 'F')) ||
-                         ((value >= 'a') && (value <= 'f'));
-
-            if (!isHex)
+            var hexValue = FromChar(value);
+            if (hexValue == 0xFF)
             {
                 return -1;
             }
 
-            if (value <= '9')
-            {
-                return value - '0';
-            }
-            else if (value <= 'F')
-            {
-                return (value - 'A') + 10;
-            }
-            else // a - f
-            {
-                return (value - 'a') + 10;
-            }
+            return hexValue;
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int FromChar(int c)
+        {
+            return c >= CharToHexLookup.Length ? 0xFF : CharToHexLookup[c];
+        }
+
+        private static ReadOnlySpan<byte> CharToHexLookup => new byte[]
```

Comment:
You could use `sbyte` instead of `byte`. That simplifies _FromChar_:

```cs
        private static int FromChar(int c)
        {
            return c >= CharToHexLookup.Length ? -1 : CharToHexLookup[c];
        }
```

And means that _ReadHex_ can end with `return FromChar(...);` with no comparison against 0xFF necessary.

Reply:
Updated to -1.